python/frequenz/microgrid_component_graph/__init__.pyi

```diff
@@ -0,0 +1,264 @@
+            disable_fallback_components: Whether to disable fallback components in
+                generated formulas.  When this is `True`, the formulas will not include
+                fallback components.
+        """
+
+ComponentT = TypeVar("ComponentT", bound=ComponentProtocol)
+ConnectionT = TypeVar("ConnectionT", bound=ConnectionProtocol)
+
+class ComponentGraph(Generic[ComponentT, ConnectionT]):
+    """A graph representation of the electrical components in a microgrid."""
+
+    def __init__(
+        self,
+        components: AbstractSet[ComponentT],
+        connections: AbstractSet[ConnectionT],
+        config: ComponentGraphConfig = ComponentGraphConfig(),
+    ) -> None:
+        """Initialize this instance.
+
+        Args:
+            components: The list of components to build the graph with.
+            connections: The list of connections between the components.
+            config: The configuration for the component graph.
+        """
+
+    def component(self, component_id: ComponentIdProtocol) -> ComponentT:
+        """Fetch the component with the specified `component_id`.
+
+        Args:
+            component_id: The id of the component to look for.
+
+        Returns:
+            The component with the given id.
+
+        Raises:
+            ValueError: if no component exists with the given ID.
+        """
+
+    def components(
+        self,
+        filter_by_ids: set[ComponentIdProtocol] | None = None,
+        filter_by_types: set[type[ComponentProtocol]] | None = None,
+    ) -> set[ComponentT]:
+        """Fetch all components in this graph.
+
+        Returns:
+            A set of all components in this graph.
+        """
+
+    def connections(self) -> set[ConnectionT]:
+        """Fetch all connections in this graph.
+
+        Returns:
+            A set of all connections in this graph.
+        """
+
+    def predecessors(self, component_id: ComponentIdProtocol) -> set[ComponentT]:
+        """Fetch all predecessors of the specified component ID.
+
+        Args:
+            component_id: ID of the component whose predecessors should be fetched.
+
+        Returns:
+            A set of components that are predecessors of the given component ID.
+
+        Raises:
+            ValueError: if no component exists with the given ID.
+        """
+
+    def successors(self, component_id: ComponentIdProtocol) -> set[ComponentT]:
+        """Fetch all successors of the specified component ID.
+
+        Args:
+            component_id: ID of the component whose successors should be fetched.
+
+        Returns:
+            A set of components that are successors of the given component ID.
+
+        Raises:
+            ValueError: if no component exists with the given ID.
+        """
+
+    def is_pv_meter(self, component_id: ComponentIdProtocol) -> bool:
+        """Check if the specified component is a PV meter.
```

Comment:
Maybe document a bit better what this means, under naive eyes this might look like this only checks what's the predecessor of a meter.